iceoryx_utils/include/iceoryx_utils/internal/relocatable_pointer/relocatable_ptr.hpp

```diff
@@ -13,188 +14,69 @@
-    {
-        return computeRawPtr();
-    }
-
-    offset_t getOffset()
-    {
-        return m_offset;
-    }
-
-
-    static constexpr offset_t NULL_POINTER_OFFSET = std::numeric_limits<offset_t>::max();
-
-  protected:
-    offset_t m_offset{NULL_POINTER_OFFSET};
-
-    inline offset_t computeOffset(const void* ptr) const
-    {
-        //@ todo: find most efficient way to do this and check the valid range (signed/unsigned issues)
-        // this implies that the absolute difference cannot be larger than 2^63 which is probably true in any shared
-        // memory we use
-        // otherwise we would need to use unsigned for differences and use one extra bit from somewhere else to indicate
-        // the sign
-
-        // this suffices if both addresses are not too far apart, e.g. when they point to data in a sufficiently "small"
-        // shared memory
-        //(if the shared memory is small, the difference does never underflow)
-
-        // todo: better first cast to unsigned, then cast to signed later (extends range where it)
-        return reinterpret_cast<offset_t>(&m_offset) - reinterpret_cast<offset_t>(ptr);
-    }
-
-    inline void* computeRawPtr() const
-    {
-        if (m_offset == NULL_POINTER_OFFSET)
-            return nullptr;
-
-        //@ todo: find most efficient way to do this (see above)
-        return reinterpret_cast<void*>(reinterpret_cast<offset_t>(&m_offset) - m_offset);
-    }
-};
-
-// typed version so we can use operator->
+/// @brief typed version so we can use operator->
 template <typename T>
-class relocatable_ptr : public RelocatablePointer
+class RelocatablePointer : public BaseRelocatablePointer
 {
   public:
-    relocatable_ptr()
-        : RelocatablePointer()
-    {
-    }
-
-    relocatable_ptr(const T* ptr)
-        : RelocatablePointer(ptr)
-    {
-    }
-
-    relocatable_ptr(const RelocatablePointer& other)
-    {
-        m_offset = computeOffset(other.computeRawPtr());
-    }
-
-    relocatable_ptr(T* rawPtr)
-    {
-        m_offset = computeOffset(rawPtr);
-    }
-
-    relocatable_ptr& operator=(const RelocatablePointer& other)
-    {
-        m_offset = computeOffset(other.computeRawPtr());
-
-        return *this;
-    }
-
-    T& operator*()
-    {
-        return *(static_cast<T*>(computeRawPtr()));
-    }
-
-    T* operator->()
-    {
-        return static_cast<T*>(computeRawPtr());
-    }
-
-    const T& operator*() const
-    {
-        return *(static_cast<T*>(computeRawPtr()));
-    }
-
-    const T* operator->() const
-    {
-        return static_cast<T*>(computeRawPtr());
-    }
-
-    T& operator[](uint64_t index)
-    {
-        auto ptr = static_cast<T*>(computeRawPtr());
-        return *(ptr + index);
-    }
-
-    operator T*() const
-    {
-        return reinterpret_cast<T*>(computeRawPtr());
-    }
+    /// @brief default constructs a logical nullptr
+    RelocatablePointer() noexcept;
+
+    /// @brief creates a RelocatablePointer pointing to ptr
```

Comment:
Depending again on the view of logical equivalence of pointers and what we want to consider the pointee.